MESOS: Switch to incremental pod config notification

sttts opened the pull request:
This fixes https://github.com/mesosphere/kubernetes-mesos/issues/463, i.e.
https://github.com/kubernetes/kubernetes/pull/13003 which broke contrib/mesos.

sttts committed:
Switch to incremental pod config notification

This fixes https://github.com/mesosphere/kubernetes-mesos/issues/463, i.e.
https://github.com/kubernetes/kubernetes/pull/13003 which broke contrib/mesos.

sttts commented:
/cc @jdef @sur for internal review

s-urbaniak commented:
LGTM

jdef commented:
I'm worried that since we no longer tell the kubelet "this is how we see the whole world" .. that things could get out of sync. Have you reviewed the kubelet code to check what else updates the kubelet state and if we need to somehow need to reconcile periodically?

sttts commented:
it's only about "our" config source. As we create it, I assume that nobody else will modify that.

sttts commented:
Though we could actually use kubelet.SET again if there is a problem like that. The main change is to switch the notification mode to incremental.

jdef commented:
Let's not assume; please check the kubelet code

On Thu, Sep 3, 2015 at 10:38 AM, Dr. Stefan Schimanski <
notifications@github.com> wrote:

> it's only about "our" config source. As we create it, I assume that nobody
> else will modify that.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/13549#issuecomment-137468492>
> .
>

jdef commented:
We can't use SET because they've hamstring'ed SET's implementation such
that it's worthless.

On Thu, Sep 3, 2015 at 10:38 AM, Dr. Stefan Schimanski <
notifications@github.com> wrote:

> Though we could actually use kubelet.SET again if there is a problem like
> that. The main change is to switch the notification mode to incremental.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/13549#issuecomment-137468712>
> .
>

sttts commented:
They have done that in the kubelet itself, not the config sources. There is still a "diff" which transforms SET to ADD+REMOVE+UPDATE.

jdef commented:
https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/kubelet.go#L1897

On Thu, Sep 3, 2015 at 10:53 AM, Dr. Stefan Schimanski <
notifications@github.com> wrote:

> They have done that in the kubelet itself, not the config sources. There
> is still a "diff" which transforms SET to ADD+REMOVE+UPDATE.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/13549#issuecomment-137475169>
> .
>

sttts commented:
It goes through the config code first: https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/config/config.go#L149

sttts commented:
And this was the SET which caused the error message: https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/config/config.go#L173

jdef commented:
lgtm

sttts commented:
There was an invalid change in the static pod test. The file source for static pods still uses kubelet.SET. I had changed the test to also expect ADD. Fixed that. Tests should be green now.

@jdef PTAL
